### PR TITLE
Implement configureFieldIndexes with tests

### PR DESCRIPTION
**Pull request by tom-andersen**

### CODE REVIEW

File: packages/firestore/src/api/credentials.ts [hunk 1]
```diff
@@ -473,24 +473,25 @@ export class FirebaseAppCheckTokenProvider
     asyncQueue: AsyncQueue,
     changeListener: CredentialChangeListener<string>
```
Comment (tom-andersen) [inline packages/firestore/src/api/credentials.ts:474]: This file is just whitespace formatting that happens when the lint:fix and yarn prettier are run.

File: packages/firestore/test/unit/local/local_store_indexeddb.test.ts [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+
+    await test.backfillIndexes();
+    await test.assertFieldsIndexes(updatedIndexA);
+
+    // Re-add the same index. We do not reset the index to its initial state.
+    await test.configureFieldsIndexes(indexA);
+    await test.assertFieldsIndexes(updatedIndexA);
+  });
+
+  it('Deleted Document Removes Index', async () => {
```
Comment (tom-andersen) [inline packages/firestore/test/unit/local/local_store_indexeddb.test.ts:319]: This test currently fails. I will need some help I think.

Comment (tom-andersen) [reply packages/firestore/test/unit/local/local_store_indexeddb.test.ts:269]: Fixed

File: packages/firestore/test/unit/local/local_store_indexeddb.test.ts [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+          doc('coll/b', 10, { count: 2 }),
+          doc('coll/c', 10, { count: 3 })
+        ],
+        [targetId]
+      )
+    );
+    await test.writeMutations(deleteMutation('coll/b'));
+    await test.backfillIndexes();
+
+    // The query engine first reads the documents by key and then re-runs the query without limit.
```
Comment (wu-hui) [inline packages/firestore/test/unit/local/local_store_indexeddb.test.ts:424]: Remove this comment.

**Review by wu-hui — COMMENTED** [review]
Mostly ready to merge. I have a kind of big ask to add a spec test..

File: packages/firestore/test/unit/local/local_store_indexeddb.test.ts (hunk 3 above)
Comment (tom-andersen) [reply packages/firestore/test/unit/local/local_store_indexeddb.test.ts]: Removed

File: packages/firestore/src/local/local_store_impl.ts [hunk 4]
```diff
@@ -1483,3 +1486,37 @@ export async function localStoreSaveNamedQuery(
...
+export async function localStoreConfigureFieldIndexes(
+  localStore: LocalStore,
+  newFieldIndexes: FieldIndex[]
+): Promise<void> {
+  const localStoreImpl = debugCast(localStore, LocalStoreImpl);
+  const indexManager = localStoreImpl.indexManager;
+  const promises: Array<PersistencePromise<void>> = [];
+  return localStoreImpl.persistence.runTransaction(
+    'Configure indexes',
+    'readwrite-primary',
```
[the remark this answers is not in the document]
Comment (tom-andersen) [reply packages/firestore/src/local/local_store_impl.ts]: Added spec test. Spec test would only pass after and changed to just 'readwrite'.

File: packages/firestore/test/unit/specs/index_spec.test.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+          }
+        ]
+      }
+    ]
+  };
+  const expectedIndexes = [
+    Helpers.fieldIndex('restaurants', {
+      fields: [['price', IndexKind.ASCENDING]]
+    })
+  ];
```
Comment (wu-hui) [inline packages/firestore/test/unit/specs/index_spec.test.ts:44]: add an empty line below.

File: packages/firestore/test/unit/specs/index_spec.test.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  specTest('Index Creation visible on all clients', ['multi-client'], () => {
+    return client(0)
+      .expectPrimaryState(true)
+      .setIndexConfiguration(config)
+      .expectIndexes(expectedIndexes)
+      .client(1)
+      .expectPrimaryState(false)
+      .expectIndexes(expectedIndexes);
+  });
```
Comment (wu-hui) [inline packages/firestore/test/unit/specs/index_spec.test.ts:53]: add an empty line below.

File: packages/firestore/src/api/credentials.ts [hunk 7]
```diff
@@ -473,24 +473,25 @@ export class FirebaseAppCheckTokenProvider
...
-      tokenResult => {
-        if (tokenResult.error != null) {
-          logDebug(
-            'FirebaseAppCheckTokenProvider',
-            `Error getting App Check token; using placeholder token instead. Error: ${tokenResult.error.message}`
-          );
-        }
-        const tokenUpdated = tokenResult.token !== this.latestAppCheckToken;
-        this.latestAppCheckToken = tokenResult.token;
+    const onTokenChanged: (
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply packages/firestore/src/api/credentials.ts:476]: This does not look like a whitespace change to me..the error message changed for example.

**Review by wu-hui — COMMENTED** [review]
Almost LGTM, I believe the app check change worth taking another look.

File: packages/firestore/src/api/credentials.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (tom-andersen) [reply packages/firestore/src/api/credentials.ts:476]: Take a closer look. Githubs diff layout is misleading you.

File: packages/firestore/test/integration/api/index_configuration.test.ts [hunk 8]
```diff
@@ -82,14 +79,12 @@ apiDescribe('Index Configuration:', (persistence: boolean) => {
 
   it('bad JSON does not crash client', () => {
     return withTestDb(persistence, async db => {
-      const action: Promise<void> = setIndexConfiguration(db, '{,}');
+      const action = (): Promise<void> => setIndexConfiguration(db, '{,}');
       if (persistence) {
-        await expect(action).to.eventually.be.rejectedWith(
-          /Failed to parse JSON/
-        );
+        expect(action).to.throw(/Failed to parse JSON/);
```
[the remark this answers is not in the document]
Comment (tom-andersen) [reply packages/firestore/test/integration/api/index_configuration.test.ts:84]: No promise is returned. An exception is immediately thrown when action is called.

File: packages/firestore/test/unit/specs/index_spec.test.ts (hunk 5 above)
Comment (tom-andersen) [reply packages/firestore/test/unit/specs/index_spec.test.ts:44]: Done

File: packages/firestore/test/unit/specs/index_spec.test.ts (hunk 6 above)
Comment (tom-andersen) [reply packages/firestore/test/unit/specs/index_spec.test.ts:53]: Done

File: packages/firestore/src/api/credentials.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply packages/firestore/src/api/credentials.ts:476]: OK, see it now!

**Review by wu-hui — APPROVED** [review]
Thanks, approved!

File: packages/firestore/package.json [hunk 9]
```diff
@@ -97,8 +97,11 @@
     "@firebase/auth": "0.20.4",
     "@rollup/plugin-alias": "3.1.9",
     "@rollup/plugin-json": "4.1.0",
+    "@types/chai": "^4.3.1",
```
Comment (hsubox76) [inline packages/firestore/package.json:100]: The repo's main package.json already has `@types/chai` and `chai` so I don't think it's necessary to add them here.

File: packages/firestore/package.json [hunk 10]
```diff
@@ -97,8 +97,11 @@
...
     "@types/eslint": "7.29.0",
     "@types/json-stable-stringify": "1.0.34",
+    "chai": "^4.3.6",
+    "chai-exclude": "^2.1.0",
```
Comment (hsubox76) [inline packages/firestore/package.json:104]: We usually pin dependencies to the specific version, so this should be `"chai-exclude": 2.1.0`

File: packages/firestore/package.json (hunk 9 above)
Comment (tom-andersen) [reply packages/firestore/package.json]: Removed.

File: packages/firestore/package.json (hunk 10 above)
Comment (tom-andersen) [reply packages/firestore/package.json]: Fixed.

**Review by hsubox76 — APPROVED** [review]

**Pull request merged**